sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/BatchGroupUtils.scala

```diff
@@ -403,23 +403,37 @@ class CombiningIterator(
   private var pendingInput: Option[SpillableColumnarBatch] = None
   Option(TaskContext.get()).foreach(onTaskCompletion(_)(pendingInput.foreach(_.close())))
 
-  // The Python output should line up row for row so we only look at the Python output
-  // iterator and no need to check the `inputPending` who will be consumed when draining
-  // the Python output.
-  override def hasNext: Boolean = pythonOutputIter.hasNext
+  private var nextReadRowsNum: Option[Int] = None
 
-  override def next(): ColumnarBatch = {
+  private def initRowsNumForNextRead(): Unit = if (nextReadRowsNum.isEmpty){
     val numRows = inputBatchQueue.peekBatchNumRows()
     // Updates the expected batch size for next read
     pythonArrowReader.setMinReadTargetNumRows(numRows)
+    nextReadRowsNum = Some(numRows)
+  }
+
+  // The Python output should line up row for row so we only look at the Python output
+  // iterator and no need to check the `pendingInput` who will be consumed when draining
+  // the Python output.
+  override def hasNext: Boolean = {
+    // pythonOutputIter.hasNext may trigger a read, so init the read rows number here.
+    initRowsNumForNextRead()
+    pythonOutputIter.hasNext
+  }
+
+  override def next(): ColumnarBatch = {
+    initRowsNumForNextRead()
```

Comment:
Why do we invoke `initRowsNumForNextRead` in both `hasNext` and `next`?
Change to? 
```
assert(nextReadRowsNum is not empty)
```

Reply:
It could be. But just in case `next` is called before `hasNext`, even it is rare.